server/src/main/java/io/druid/server/security/Authenticator.java

```diff
@@ -0,0 +1,107 @@
+    @JsonSubTypes.Type(name = "noop", value = NoopAuthenticator.class),
+})
+/**
+ * This interface is essentially a ServletFilterHolder with additional requirements on the getFilter() method contract, plus:
+ *
+ * - A method that returns a WWW-Authenticate challenge header appropriate for the authentication mechanism.
+ * - A method for creating a wrapped HTTP client that can authenticate using the Authenticator's authentication scheme,
+ *   used for internal Druid node communications (e.g., broker -> historical messages)
+ * - A method for authenticating credentials contained in a JDBC connection context, used for authenticating Druid SQL
+ *   requests received via JDBC
+ */
+public interface Authenticator extends ServletFilterHolder
+{
+  /**
+   * @return The namespace associated with this Authenticator. This will be used for choosing the correct
+   * Authorizer for authorizing requests that have been authenticated by this Authenticator.
+   */
+  public String getNamespace();
+
+  /**
+   * Create a Filter that performs authentication checks on incoming HTTP requests.
+   * <p>
+   * If the authentication succeeds, the Filter should set the "Druid-Auth-Token" attribute in the request,
+   * containing a String that represents the authenticated identity of the requester.
+   * <p>
+   * If the "Druid-Auth-Token" attribute is already set (i.e., request has been authenticated by an earlier Filter),
+   * this Filter should skip any authentication checks and proceed to the next Filter.
+   * <p>
+   * If the authentication fails, the Filter should not send an error response. The error response will be sent
+   * after all Filters in the authentication filter chain have been checked.
+   * <p>
+   * If an anonymous request is received, the Filter should continue on to the next Filter, the challenge response
+   * will be sent after the filter chain is exhausted.
+   *
+   * @return Filter that authenticates HTTP requests
+   */
+  @Override
+  public Filter getFilter();
+
+  /**
+   * Return a WWW-Authenticate challenge scheme string appropriate for this Authenticator's authentication mechanism.
+   * <p>
+   * For example, a Basic HTTP implementation should return "Basic", while a Kerberos implementation would return
+   * "Negotiate".
+   *
+   * @return Authentication scheme
+   */
+  public String getAuthChallengeHeader();
+
+  /**
+   * Given a JDBC connection context, authenticate the identity represented by the information in the context.
+   * This is used to secure JDBC access for Druid SQL.
+   * <p>
+   * For example, a Basic HTTP auth implementation could read the "user" and "password" fields from the JDBC context.
+   * <p>
+   * The expected contents of the context are left to the implementation.
+   *
+   * @param context JDBC connection context
+   *
+   * @return true if the identity represented by the context is successfully authenticated
+   */
+  public boolean authenticateJDBCContext(Map<String, Object> context);
```

Comment:
shouldn't this return String Authorizer name so that appropriate Authorizer instance could be obtained and authorization is done?

Reply:
Changed this to return a nullable AuthenticationResult now, which contains identity and Authorizer name